File: geneintersection/python/generegion.py
#!/usr/bin/env python
# coding=utf-8
"""
intersection > THRESHOLD

Usage:
THRESHOLD:
    python argv[0] > outputfile
RATIO
    python runratio.py

It takes ~20-30s to run.

"""
import sys

import numpy as np
import pandas

# sys.path.append("./")

THRESHOLD = 50
RATIO = float(sys.argv[1])
SIMILAR_NEIGHBOR = 3

data = pandas.read_csv("./python_test_data.csv")

data = data[0:935]
data = data.fillna(0)

dataSE = pandas.DataFrame(
    data,
    columns=[
        "sample1_Pos1", "sample1_Pos2", "sample2_Pos1", "sample2_Pos2",
        "sample3_Pos1", "sample3_Pos2"
    ])
dataSE = dataSE.astype(int)

Sam1S = np.array(dataSE["sample1_Pos1"], dtype=int)
Sam1E = np.array(dataSE["sample1_Pos2"], dtype=int)
Sam2S = np.array(dataSE["sample2_Pos1"], dtype=int)
Sam2E = np.array(dataSE["sample2_Pos2"], dtype=int)
Sam3S = np.array(dataSE["sample3_Pos1"], dtype=int)
Sam3E = np.array(dataSE["sample3_Pos2"], dtype=int)

metadata = list()


def extractRegion(Sample1, Sample2, Sample3):
    for i in range(len(Sample1)):
        for j in range(len(Sample2)):

            if data["sample1_Chr1"][i] != data["sample2_Chr1"][j]:
                continue

            if min(Sam1E[i], Sam2E[j]) < max(Sam1S[i], Sam2S[j]):
                continue

            for k in range(len(Sample3)):

                # Some CHR
                if data["sample1_Chr1"][i] != data["sample3_Chr1"][k] or\
                   data["sample2_Chr1"][j] != data["sample3_Chr1"][k]:
                    continue

                if min(Sam1E[i], Sam3E[k]) < max(Sam1S[i], Sam3S[k]) or \
                        min(Sam2E[j], Sam3E[k]) < max(Sam2S[j], Sam3S[k]):
                    continue
                else:
                    maxS = np.max((Sam1S[i], Sam2S[j], Sam3S[k]), axis=0)
                    minE = np.min((Sam1E[i], Sam2E[j], Sam3E[k]), axis=0)

                    # if (minE - maxS) > THRESHOLD:
                    intersect_length = minE - maxS
                    sample1_length = Sam1E[i] - Sam1S[i]
                    sample2_length = Sam2E[j] - Sam2S[j]
                    sample3_length = Sam3E[k] - Sam3S[k]
                    sample1_ratio = float(intersect_length) / sample1_length
                    sample2_ratio = float(intersect_length) / sample2_length
                    sample3_ratio = float(intersect_length) / sample3_length
                    if sample1_ratio == 1.0 and min(sample1_ratio,
                                                    sample2_ratio,
                                                    sample3_ratio) > RATIO:
                        metadata.append([
                            data["sample1_Chr1"][i], Sam1S[i], Sam1E[i],
                            data["sample2_Chr1"][j], Sam2S[j], Sam2E[j],
                            data["sample3_Chr1"][k], Sam3S[k], Sam3E[k],
                            Sam1E[i] - Sam1S[i], Sam2E[j] - Sam2S[j],
                            Sam3E[k] - Sam3S[k], sample1_ratio, sample2_ratio,
                            sample3_ratio
                        ])


header = [
    "sample1_Chr1",
    "sample1_start(Pos1)",
    "sample1_end(Pos2)",
    "sample2_Chr1",
    "sample2_start(Pos1)",
    "sample2_end(Pos2)",
    "sample3_Chr1",
    "sample3_start(Pos1)",
    "sample3_end(Pos2)",
    "sample1 region length",
    "sample2 region length",
    "sample3 region length",
    "ratio of sample1",
    "ratio of sample2",
    "ratio of sample3",
]

extractRegion(Sam1S, Sam2S, Sam3S)

print(header)

for i in metadata:
    print(i)
# metadata = np.array(metadata)
# metadata = pandas.DataFrame(metadata)
# metadata.to_csv("intersectRegionResult.csv", header=header, index=False)

File: geneintersection/cpp/generegion.cpp
/**
 * ./GeneRegion v0.1
 * 2017年 12月 04日 星期一 16:42:27 CST
 *
 * Usage:./GeneRegion > result.csv
 * TODO:
 * 1.Read the whole datafile rather than a certain number of lines.
 * 2.Parse arguments.
 * 3.Use a better strategy to find similar genes, such as a good distance.
 *
 */
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unistd.h>
#include <vector>

using namespace std;

/* Parameters:
 * LENGTHOFFILE:
 *      The last part data  of python_test_data is not integrated.
 * RATIO:
 *      Min Ratio of intersect length.
 * DISTANCETHRESHOLD:
 *      The distance of three gene region SHOULD be small.
 *
*/
const int LENGTHOFFILE = 870;
const float RATIO = 0.9;
const int DISTANCETHRESHOLD = 5000;

// Nearest N Neighbor strategy.
// const int SIMILAR_NEIGHBOR = 3;

string str_header[] = {
    "sample1_Chr1",          "sample1_start(Pos1)",   "sample1_end(Pos2)",
    "sample2_Chr1",          "sample2_start(Pos1)",   "sample2_end(Pos2)",
    "sample3_Chr1",          "sample3_start(Pos1)",   "sample3_end(Pos2)",
    "sample1 region length", "sample2 region length", "sample3 region length",
    "ratio of sample1",      "ratio of sample2",      "ratio of sample3"};

vector<string> sample1_chromosome_name;
vector<int> sample1_region_start;
vector<int> sample1_region_end;

vector<string> sample2_chromosome_name;
vector<int> sample2_region_start;
vector<int> sample2_region_end;

vector<string> sample3_chromosome_name;
vector<int> sample3_region_start;
vector<int> sample3_region_end;

vector<string> header(str_header,
                      str_header + int(sizeof(str_header) / sizeof(string)));

void print_header();
void print_help(int argc, char *argv[]);
int read_csvfile(char *filepath);
float distance_three_region(int, int, int, int, int, int);

int main(int argc, char *argv[]) {

  /* TODO parse argument
  if ( argc == 1 ){
      print_help(argc, argv);
      exit(1);
  }
  */

  // TODO:check the existence of datafile.
  // read_csvfile((char*)argv[1]);
  read_csvfile((char *)"./python_test_data.csv");
  print_header();

  for (int i = 0; i < LENGTHOFFILE; ++i) {
    for (int j = 0; j < LENGTHOFFILE; ++j) {

      if (sample1_chromosome_name[i] != sample2_chromosome_name[j]) {
        continue;
      }
      if (min(sample1_region_end[i], sample2_region_end[j]) <
          max(sample1_region_start[i], sample2_region_start[j])) {
        continue;
      }

      for (int k = 0; k < LENGTHOFFILE; ++k) {

        if (sample1_chromosome_name[i] != sample3_chromosome_name[k] &&
            sample2_chromosome_name[j] != sample3_chromosome_name[k]) {
          continue;
        }
        if (min(sample1_region_end[i], sample3_region_end[k]) <
            max(sample2_region_start[j], sample3_region_start[k])) {
          continue;
        }
        int max_start =
            max(max(sample1_region_start[i], sample2_region_start[j]),
                sample3_region_start[k]);
        int min_end = min(min(sample1_region_end[i], sample2_region_end[j]),
                          sample3_region_end[k]);
        int intersect_length = min_end - max_start;
        int sample1_region_length =
            sample1_region_end[i] - sample1_region_start[i];
        int sample2_region_length =
            sample2_region_end[j] - sample2_region_start[j];
        int sample3_region_length =
            sample3_region_end[k] - sample3_region_start[k];
        float sample1_ratio = float(intersect_length) / sample1_region_length;
        float sample2_ratio = float(intersect_length) / sample2_region_length;
        float sample3_ratio = float(intersect_length) / sample3_region_length;

        if (distance_three_region(
                sample1_region_start[i], sample1_region_end[i],
                sample2_region_start[j], sample2_region_end[j],
                sample3_region_start[k],
                sample3_region_end[k]) <= DISTANCETHRESHOLD)
          if (min(min(sample1_ratio, sample2_ratio), sample3_ratio) > RATIO) {
            cout << sample1_chromosome_name[i] << ',' << sample1_region_start[i]
                 << ',' << sample1_region_end[i] << ','

                 << sample2_chromosome_name[j] << ',' << sample2_region_start[j]
                 << ',' << sample2_region_end[j] << ','

                 << sample3_chromosome_name[k] << "," << sample3_region_start[k]
                 << ',' << sample3_region_end[k] << ','

                 << sample1_region_length << ',' << sample2_region_length << ','
                 << sample3_region_length << ','

                 << sample1_ratio << ',' << sample2_ratio << ','
                 << sample3_ratio << endl;
            // cout << "(i, j, k):(" << i << "," << j << "," << k  << ")"<<
            // endl;
            /*
            if (sample1_chromosome_name[i] == "X1"){
                cout << sample1_region_start[i] << " " <<  sample1_region_end[i]
            << endl;
            }
            */
          }
      }
    }
  }

  return 0;
}

void print_header() {
  for (vector<string>::iterator it = header.begin(); it != header.end(); ++it) {
    cout << *it << ',';
  }
  cout << endl;

  return;
}

void print_help(int argc, char *argv[]) {
  cout << "argc: " << argc << endl;
  cout << "Usage: " << argv[0] << "  "
       << "datafile.csv" << endl;
}

int read_csvfile(char *filepath) {
  ifstream ifs(filepath);
  string orgin_header;
  getline(ifs, orgin_header);
  for (int i = 0; i < LENGTHOFFILE; ++i) {

    string chromosome_name;
    int region_start;
    int region_end;

    getline(ifs, chromosome_name, ',');
    ifs >> region_start;
    ifs.get();
    getline(ifs, chromosome_name, ',');
    ifs >> region_end;
    ifs.get();
    sample1_chromosome_name.push_back(chromosome_name);
    sample1_region_start.push_back(region_start);
    sample1_region_end.push_back(region_end);

    getline(ifs, chromosome_name, ',');
    ifs >> region_start;
    ifs.get();
    getline(ifs, chromosome_name, ',');
    ifs >> region_end;
    ifs.get();
    sample2_chromosome_name.push_back(chromosome_name);
    sample2_region_start.push_back(region_start);
    sample2_region_end.push_back(region_end);

    getline(ifs, chromosome_name, ',');
    ifs >> region_start;
    ifs.get();
    getline(ifs, chromosome_name, ',');
    ifs >> region_end;
    ifs.get();
    sample3_chromosome_name.push_back(chromosome_name);
    sample3_region_start.push_back(region_start);
    sample3_region_end.push_back(region_end);
  }

  return 1;
}

float distance_three_region(int s1, int e1, int s2, int e2, int s3, int e3) {
  float d12 = abs(s1 - s2) + abs(e1 - e2);
  float d13 = abs(s1 - s3) + abs(e1 - e3);
  float d23 = abs(s2 - s3) + abs(e2 - e3);

  return d12 + d13 + d23;
}
